model/project.go

```diff
@@ -1631,6 +1633,115 @@ func (p *Project) GetDisplayTask(variant, name string) *patch.DisplayTask {
 	return nil
 }
 
+func GetVersionsWithOptions(projectId string, opts GetVersionsOptions) ([]Version, error) {
+	match := bson.M{
+		VersionIdentifierKey: projectId,
+	}
+	if opts.ByBuildVariant != "" {
+		match[bsonutil.GetDottedKeyName(VersionBuildVariantsKey, VersionBuildStatusVariantKey)] = opts.ByBuildVariant
+	}
+	if opts.VersionToStartAt != "" {
+		v, err := VersionFindOneId(opts.VersionToStartAt)
+		if err != nil {
+			return nil, errors.Wrapf(err, "problem finding version '%s'", opts.VersionToStartAt)
+		}
+		if v == nil {
+			return nil, errors.Errorf("version '%s' doesn't exist", opts.VersionToStartAt)
+		}
+		match[VersionCreateTimeKey] = bson.M{"$lte": v.CreateTime}
+	}
+
+	pipeline := []bson.M{bson.M{"$match": match}}
+	pipeline = append(pipeline, bson.M{"$sort": bson.M{VersionCreateTimeKey: -1}})
+	project := bson.M{
+		VersionRevisionKey:            1,
+		VersionErrorsKey:              1,
+		VersionMessageKey:             1,
+		VersionAuthorKey:              1,
+		VersionRevisionOrderNumberKey: 1,
+		VersionCreateTimeKey:          1,
+		VersionStartTimeKey:           1,
+		VersionFinishTimeKey:          1,
+		VersionStatusKey:              1,
+		VersionBuildVariantsKey:       1,
+	}
+
+	if opts.IncludeBuilds {
+		if opts.ByBuildVariant != "" {
+			// add a filter so we only have the build variant we want (we already know this exists from the match stage)
+			project[VersionBuildVariantsKey] = bson.M{
+				"$filter": bson.M{
+					"input": "$" + VersionBuildVariantsKey,
+					"as":    "item",
+					"cond": bson.M{
+						"$eq": []string{"$$item.build_variant", opts.ByBuildVariant},
+					},
+				},
+			}
+		}
+
+		pipeline = append(pipeline, bson.M{"$project": project})
+		lookupBuilds := bson.M{
+			"from":         build.Collection,
+			"localField":   bsonutil.GetDottedKeyName(VersionBuildVariantsKey, VersionBuildStatusBuildIdKey),
+			"foreignField": build.IdKey,
+			"as":           "build_variants",
+		}
+		pipeline = append(pipeline, bson.M{"$lookup": lookupBuilds})
+		// projectWithBuilds is initially the same as the first project but with build_variants instead of build_variants_status.
+		projectWithBuilds := bson.M{
+			VersionRevisionKey:            1,
+			VersionErrorsKey:              1,
+			VersionMessageKey:             1,
+			VersionAuthorKey:              1,
+			VersionRevisionOrderNumberKey: 1,
+			VersionCreateTimeKey:          1,
+			VersionStartTimeKey:           1,
+			VersionFinishTimeKey:          1,
+			VersionStatusKey:              1,
+			"build_variants":              1,
+		}
+		pipeline = append(pipeline, bson.M{"$project": projectWithBuilds})
+	}
+	if opts.Skip != 0 {
+		pipeline = append(pipeline, bson.M{"$skip": opts.Skip})
+	}
+	if opts.Limit != 0 {
+		pipeline = append(pipeline, bson.M{"$limit": opts.Limit})
```

Comment:
I would put the default limit here rather than in the route. That way if something else ends up calling this they won't accidentally return all versions in a project